fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/impl/object/builder/FabricTrackedDataRegistryImpl.java

```diff
@@ -0,0 +1,127 @@
+	private static List<TrackedDataHandler<?>> EXTERNAL_MODDED_HANDLERS = new ArrayList<>();
+
+	private FabricTrackedDataRegistryImpl() {
+	}
+
+	public static void storeVanillaHandlers() {
+		Int2ObjectBiMap<TrackedDataHandler<?>> dataHandlers = TrackedDataHandlerRegistryAccessor.fabric_getDataHandlers();
+
+		for (TrackedDataHandler<?> handler : dataHandlers) {
+			VANILLA_HANDLERS.add(handler);
+		}
+
+		LOGGER.debug("Stored {} vanilla handlers", VANILLA_HANDLERS.size());
+	}
+
+	/**
+	 * Reorders handlers in {@link net.minecraft.entity.data.TrackedDataHandlerRegistry#DATA_HANDLERS} to have a consistent order between client and server.
+	 *
+	 * <p>The order used is the following:
+	 *
+	 * <ul>
+	 *   <li>Vanilla handlers</li>
+	 *   <li>Handlers in the Fabric API registry (sorted by ID)</li>
+	 *   <li>External modded handlers</li>
+	 * </ul>
+	*/
+	private static void reorderHandlers() {
+		Int2ObjectBiMap<TrackedDataHandler<?>> dataHandlers = TrackedDataHandlerRegistryAccessor.fabric_getDataHandlers();
+		LOGGER.debug("Reordering tracked data handlers containing {} entries", dataHandlers.size());
+
+		// Store external modded handlers
+		for (TrackedDataHandler<?> handler : dataHandlers) {
+			if (VANILLA_HANDLERS.contains(handler)) continue;
+			if (HANDLER_REGISTRY != null && HANDLER_REGISTRY.getId(handler) != null) continue;
+			if (EXTERNAL_MODDED_HANDLERS.contains(handler)) continue;
+
+			EXTERNAL_MODDED_HANDLERS.add(handler);
+			LOGGER.warn("Tracked data handler {} is not managed by vanilla or Fabric API; it may be prone to desynchronization!", handler);
+		}
+
+		// Reset the map so that handlers can be added back in a new order
+		dataHandlers.clear();
+
+		// Add handlers back to map
+		for (TrackedDataHandler<?> handler : VANILLA_HANDLERS) {
+			dataHandlers.add(handler);
+		}
+
+		if (HANDLER_REGISTRY != null) {
+			for (TrackedDataHandler<?> handler : HANDLER_REGISTRY) {
+				dataHandlers.add(handler);
+			}
+		}
+
+		for (TrackedDataHandler<?> handler : EXTERNAL_MODDED_HANDLERS) {
+			dataHandlers.add(handler);
+		}
+
+		LOGGER.debug("Finished reordering tracked data handlers containing {} entries", dataHandlers.size());
+	}
+
+	public static void registerHandler(Identifier id, TrackedDataHandler<?> handler) {
+		Objects.requireNonNull(id, "Tracked data handler ID cannot be null!");
+		Objects.requireNonNull(handler, "Tracked data handler cannot be null!");
+
+		if (VANILLA_HANDLERS.contains(handler) || EXTERNAL_MODDED_HANDLERS.contains(handler)) {
+			throw new IllegalArgumentException("Cannot register tracked data handler added via TrackedDataHandlerRegistry.register");
+		}
+
+		if (HANDLER_REGISTRY == null) {
+			HANDLER_REGISTRY = FabricRegistryBuilder
+					.createSimple(HANDLER_REGISTRY_KEY)
+					.attribute(RegistryAttribute.SYNCED)
+					.buildAndRegister();
+
+			RegistryIdRemapCallback.event(HANDLER_REGISTRY).register(state -> reorderHandlers());
```

Comment:
Would Client/ServerPlayConnectionEvents.JOIN be a better place to do this? I dont think it would be a bad idea to sort them event when a modded one isnt registered.